packit/sync.py

```diff
@@ -1,127 +1,144 @@
-# MIT License
-#
-# Copyright (c) 2019 Red Hat, Inc.
-
-# Permission is hereby granted, free of charge, to any person obtaining a copy
-# of this software and associated documentation files (the "Software"), to deal
-# in the Software without restriction, including without limitation the rights
-# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-# copies of the Software, and to permit persons to whom the Software is
-# furnished to do so, subject to the following conditions:
-#
-# The above copyright notice and this permission notice shall be included in all
-# copies or substantial portions of the Software.
-#
-# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-# SOFTWARE.
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
 
+import glob
 import logging
-import shutil
 from pathlib import Path
-from typing import List
-from typing import NamedTuple, Union
+from typing import List, Optional, Union, Sequence
 
 from packit.exceptions import PackitException
+from packit.utils import run_command
 
 logger = logging.getLogger(__name__)
 
 
-class SyncFilesItem(NamedTuple):
-    src: Union[str, List[str]]
-    dest: str
+def check_subpath(subpath: Path, path: Path) -> Path:
+    """Check if 'subpath' is a subpath of 'path'
+    Args:
+        subpath: Subpath to be checked.
+        path: Path agains which subpath is checked.
+    Returns:
+        'subpath', in case it is a subpath of 'path'.
+    Raises:
+        PackitException, if 'subpath' is not a subpath of 'path'.
+    """
+    if not subpath.resolve().is_relative_to(path.resolve()):
+        raise PackitException(
+            f"Sync files: Illegal path! {subpath} is not in the subpath of {path}."
+        )
+    return subpath
 
-    def __repr__(self):
-        return f"SyncFilesItem(src={self.src}, dest={self.dest})"
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, SyncFilesItem):
-            raise NotImplementedError()
+class SyncFilesItem:
+    """Some files to sync to destination
 
-        return self.src == other.src and self.dest == other.dest
+    Think about this as a wrapper around 'rsync'.
 
+    Attributes:
+        src: List of paths to sync.
+        dest: Destination to sync to.
+        mkpath: Create the destination's path component.
```

Comment:
This is introduced in order to match the functionality of the old implementation, although strictly speaking the change is not backward compatible.

The previous implementation did create intermediary paths in dist-git if they were missing.

I've checked the configuration of our most prominent users and none of them is syncing files, where creating intermediary paths would be required. So hopefully not having this on by default, will not hurt anyone.

Examples:

This would work:

```yaml
synced_files:
- fedora/
- some.spec
```

This wouldn't:

```yaml
synced_files:
- src: fedora/some.spec
  dest: potentially/missing/intermediary/path/some.spec
```

But can be fixed with:

```yaml
synced_files:
- src: fedora/some.spec
  dest: potentially/missing/intermediary/path/some.spec
  mkpath: true
```